llvm/lib/CAS/CASDB.cpp

```diff
@@ -96,6 +97,22 @@ void HierarchicalTreeBuilder::pushImpl(Optional<CASID> ID,
   Entries.emplace_back(ID, Kind, canonicalize(CanonicalPath, Kind));
 }
 
+Error HierarchicalTreeBuilder::pushDirectory(CASDB &CAS, CASID ID,
+                                             const Twine &Path) {
+  // FIXME: Look into a more efficient way to merge directory contents.
+  return walkFileTreeRecursively(
```

Comment:
This should be hollowed out, with the logic incorporated into `create()`.

I think an efficient algorithm would be:
1. Change the sort so that for equal paths, you get `pushDirectory` then `pushTreeContent` then `push` calls.
2. Add some accounting of "live" `pushTreeContent` calls.
3. Change the `for` loop to splice in the tree content, only expanding down when/where necessary.

Another / similar option would be to do the splicing as an initial pass (after sorting) and leave the existing `for` loop as is.

For now, I suggest adding a FIXME complaining about the unnecessary work that expands the whole recursive tree (maybe describing an efficient algorithm), and then do something like this (BEFORE the sort):
 ```
// Expand out tree content now that we have a CAS.
//
// FIXME: ...
if (HasAnyTreeContent) {
  for (size_t I = 0, IE = Entries.size(); I != IE Entry; ++I) {
    if (!Entries[I].isTreeContent())
      continue;

    // Move out.
    HierchicalTreeEntry Entry = std::move(Entries[I]);

    // Expand content using `push` and `pushDirectory()` calls.
    walkFileTreeRecursively(/* ... */);

    // Leave behind a redundant pushDirectory() equivalent.
    Entries[I] = HierarchicalEntry(None, Tree, Entry.getPath());
  }
}
```

Reply:
An alternative / simpler approach (at least for the short-term solution) would be to leave `HierarchicalEntry` unchanged and add a separate vector to the builder:
```
SmallVector<HierarchicalEntry, 8> Entries;
SmallVector<HierarchicalEntry, 0> TreeContent;
```
At the top of `create()` you do a walk through `TreeContent` expanding now that you have a CAS. And add the same FIXME about splicing the content in somehow without doing a full recursive listing.

Reply:
Or, if maintaining order of entries/pushes is important (probably it is...?), do the first approach, but, instead of modifying the existing `Entries` vector, create a new one (and swap with the old after building it) that expands the tree contents during iteration of the old one.